ros2-library/src/main/java/us/ihmc/ros2/ROS2NodeBuilder.java

```diff
@@ -0,0 +1,438 @@
+      INTRAPROCESS_ONLY
+   }
+
+   private int domainId = UNSET_DOMAIN_ID;
+   private String namespace = DEFAULT_NAMESPACE;
+   private boolean useSharedMemory = true;
+   private InetAddress[] addressRestriction = null;
+   private boolean parseEnvironment = true;
+   private boolean parseProperties = true;
+   private boolean parseNetworkParametersConfig = true;
+   @Nullable
+   private SpecialTransportMode specialTransportMode;
+
+   private final transient StringJoiner buildPrintout = new StringJoiner("\n\t\t");
+
+   public ROS2NodeBuilder domainId(int domainId)
+   {
+      this.domainId = domainId;
+      return this;
+   }
+
+   public ROS2NodeBuilder namespace(String namespace)
+   {
+      this.namespace = namespace;
+      return this;
+   }
+
+   public ROS2NodeBuilder useSharedMemory(boolean useSharedMemory)
+   {
+      this.useSharedMemory = useSharedMemory;
+      return this;
+   }
+
+   public ROS2NodeBuilder addressRestriction(InetAddress... addressRestriction)
+   {
+      this.addressRestriction = addressRestriction;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseEnvironment(boolean parseEnvironment)
+   {
+      this.parseEnvironment = parseEnvironment;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseProperties(boolean parseProperties)
+   {
+      this.parseProperties = parseProperties;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseNetworkParametersConfig(boolean parseNetworkParametersConfig)
+   {
+      this.parseNetworkParametersConfig = parseNetworkParametersConfig;
+      return this;
+   }
+
+   public ROS2NodeBuilder specialTransportMode(@Nullable SpecialTransportMode specialTransportMode)
+   {
+      this.specialTransportMode = specialTransportMode;
+      return this;
+   }
+
+   public ROS2Node build(String name)
+   {
+      buildPrintout.add("Building ROS2Node: " + name);
+      return new ROS2Node(name, namespace, buildProfile(), specialTransportMode);
+   }
+
+   public RealtimeROS2Node buildRealtime(String name)
+   {
+      return buildRealtime(name, new PeriodicNonRealtimeThreadSchedulerFactory());
+   }
+
+   public RealtimeROS2Node buildRealtime(String name, PeriodicThreadSchedulerFactory threadFactory)
+   {
+      buildPrintout.add("Building RealtimeROS2Node: " + name);
+      return new RealtimeROS2Node(name, namespace, buildProfile(), specialTransportMode, threadFactory);
+   }
+
+   private ParticipantProfile buildProfile()
+   {
+      ParticipantProfile profile = ParticipantProfile.create();
+
+      // Set up ROS Domain ID
+      {
+         if (domainIDValid(domainId))
+         {
+            buildPrintout.add("Using a programmatically set ROS Domain ID: " + domainId);
+         }
+         else
+         {
+            // Try to find a ROS Domain ID
+            domainId = findDomainID();
+
+            // If a valid domain ID was not found automatically
+            if (!domainIDValid(domainId))
+            {
+               domainId = 100;
```

Comment:
Maybe get a random value?

Reply:
I wanted to keep behavior repeatable and not introduce randomness